RecoParticleFlow/PFProducer/plugins/MLPFProducer.cc

```diff
@@ -0,0 +1,137 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+#include "RecoParticleFlow/PFProducer/interface/MLPFModel.h"
+
+struct MLPFCache {
+  std::atomic<tensorflow::GraphDef*> graph_def;
+};
+
+class MLPFProducer : public edm::stream::EDProducer<edm::GlobalCache<MLPFCache> > {
+public:
+  explicit MLPFProducer(const edm::ParameterSet&, const MLPFCache*);
+  void produce(edm::Event& event, const edm::EventSetup& setup) override;
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  // static methods for handling the global cache
+  static std::unique_ptr<MLPFCache> initializeGlobalCache(const edm::ParameterSet&);
+  static void globalEndJob(MLPFCache*);
+
+private:
+  const edm::EDPutTokenT<reco::PFCandidateCollection> pfCandidatesToken_;
+  const edm::EDGetTokenT<reco::PFBlockCollection> inputTagBlocks_;
+  const std::string model_path_;
+  tensorflow::Session* session_;
+};
+
+MLPFProducer::MLPFProducer(const edm::ParameterSet& cfg, const MLPFCache* cache)
+    : pfCandidatesToken_{produces<reco::PFCandidateCollection>()},
```

Comment:
```suggestion
    : pfCandidatesPutToken_{produces<reco::PFCandidateCollection>()},
```
just to be more explicit that this is an output

Reply:
good point, changed here and elsewhere. 486bbcbc2d3